v2/internal/reconcilers/entra/entra_securitygroup_reconciler.go

```diff
@@ -0,0 +1,499 @@
+	}
+
+	err = client.Client().Groups().ByGroupId(id).Delete(ctx, nil)
+	if err != nil {
+		// If the group doesn't exist, return nil and nil as we've successfully ensured that it doesn't exist
+		if r.isNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		// If the error is not a 404, return the error
+		var displayName string
+		if group.Spec.DisplayName != nil {
+			displayName = *group.Spec.DisplayName
+		}
+
+		return ctrl.Result{}, eris.Wrapf(err, "failed to delete group %q (%s)", displayName, id)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *EntraSecurityGroupReconciler) Claim(
+	ctx context.Context,
+	log logr.Logger,
+	eventRecorder record.EventRecorder,
+	obj genruntime.MetaObject,
+) error {
+	// Nothing to do
+	//!! Confirm this is true
+	return nil
+}
+
+// update completes our reconciliation by updating an existing Entra security group.
+// ctx is the context for the operation.
+// id is the Entra ID of the group to update.
+// group is the security group to update.
+// eventRecorder is used to record events for the group.
+// log is the logger to use for logging.
+func (r *EntraSecurityGroupReconciler) update(
+	ctx context.Context,
+	id string,
+	group *asoentra.SecurityGroup,
+	log logr.Logger,
+) (ctrl.Result, error) {
+	log = log.WithValues("id", id)
+	log.V(Status).Info("Updating Entra security group")
+
+	// Create our Entra Client
+	client, err := r.EntraClientFactory(ctx, group)
+	if err != nil {
+		return ctrl.Result{}, eris.Wrap(err, "creating entra client prior to update")
+	}
+
+	// Load the existing group by ID
+	g, err := r.loadGroupByID(ctx, id, client.Client())
+	if err != nil {
+		if r.isNotFound(err) {
+			// Group used to exist, but no longer does - it's probably been deleted
+			// Remove the existing annotation and requeue the reconciliation to create a replacement
+			log.V(Status).Info("Group no longer exists")
+			setEntraID(group, "")
+			return ctrl.Result{
+				Requeue: true,
+			}, nil
+		}
+
+		return ctrl.Result{}, eris.Wrapf(err, "getting group by ID %s", id)
+	}
+
+	// Update - PATCH
+	group.Spec.AssignToGroup(g)
+
+	_, err = client.Client().Groups().ByGroupId(id).Patch(ctx, g, nil)
+	if err != nil {
+		// Failed to update
+		return ctrl.Result{}, eris.Wrapf(err, "failed to update group %s", id)
+	}
+
+	group.Status.AssignFromGroup(g)
+
+	return ctrl.Result{}, nil
+}
+
+// tryAdopt tries to find an existing Entra security group to adopt.
+// ctx is the context for the operation.
+// id is the Entra ID of the group to update.
+// group is the security group to update.
+// eventRecorder is used to record events for the group.
+// log is the logger to use for logging.
+// Returns the Entra ID of the group to adopt, and nil if found;
+// nil, nil if no group was found to adopt;
+// or nil, error if there was a problem finding the group.
+func (r *EntraSecurityGroupReconciler) tryAdopt(
+	ctx context.Context,
+	group *asoentra.SecurityGroup,
+	log logr.Logger,
+) (*string, error) {
```

Comment:
minor, style: could avoid str-ptr return here probably if you wanted and just return empty string if the ID wasn't found?

Reply:
Good idea.